ingester/ingester_lifecycle.go

```diff
@@ -0,0 +1,419 @@
+// Responsible for managing the ingester lifecycle.
+
+package ingester
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+
+	"github.com/weaveworks/common/user"
+	"github.com/weaveworks/cortex"
+	"github.com/weaveworks/cortex/ingester/client"
+	"github.com/weaveworks/cortex/ring"
+	"github.com/weaveworks/cortex/util"
+)
+
+const (
+	infName            = "eth0"
+	claimRemoteTimeout = 10 * time.Second
+	minReadyDuration   = 1 * time.Minute
+)
+
+var (
+	consulHeartbeats = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "cortex_ingester_consul_heartbeats_total",
+		Help: "The total number of heartbeats sent to consul.",
+	})
+)
+
+func init() {
+	prometheus.MustRegister(consulHeartbeats)
+}
+
+// ReadinessHandler is used to indicate to k8s when the ingesters are ready for
+// the addition removal of another ingester. Returns 204 when the ingester is
+// ready, 500 otherwise.
+func (i *Ingester) ReadinessHandler(w http.ResponseWriter, r *http.Request) {
+	if i.isReady() {
+		w.WriteHeader(http.StatusNoContent)
+	} else {
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+}
+
+func (i *Ingester) isReady() bool {
+	i.readyLock.Lock()
+	defer i.readyLock.Unlock()
+
+	if i.ready {
+		return true
+	}
+
+	// Ingester always take at least minReadyDuration to become ready to work
+	// around race conditions with ingesters exiting and updating the ring
+	if time.Now().Sub(i.startTime) < minReadyDuration {
+		return false
+	}
+
+	ringDesc, err := i.consul.Get(ring.ConsulKey)
+	if err != nil {
+		log.Error("Error talking to consul: %v", err)
+		return false
+	}
+
+	i.ready = i.ready || ringDesc.(*ring.Desc).Ready(i.cfg.ringConfig.HeartbeatTimeout)
+	return i.ready
+}
+
+// ChangeState of the ingester, for use off the loop() goroutine.
+func (i *Ingester) ChangeState(state ring.IngesterState) error {
+	err := make(chan error)
+	i.actorChan <- func() {
```

Comment:
Sorry, I should say this where it's defined, but I don't think `actorChan` is a great name. It describes an implementation strategy rather than a purpose.